syntax = "proto3";

message TupleType{
	string type1_name=1;
	string type2_name=2;
}

// message Tuple{
// 	TupleType type=1;
// 	optional NumpyNdArray numpy_array=2;
// 	optional bytes bytes_array=3;
// 	optional double double_array=4;
// 	optional float float_array=5;
// 	optional int32 int32_array=6;
// 	optional int64 int64_array=7;
// 	optional uint32 uint32_array=8;
// 	optional uint64 uint64_array=9;
// 	optional sint32 sint32_array=10;
// 	optional sint64 sint64_array=11;
// 	optional fixed32 fixed32_array=12;
// 	optional fixed64 fixed64_array=13;
// 	optional sfixed32 sfixed32_array=14;
// 	optional sfixed64 sfixed64_array=15;
// 	optional bool bool_array=16;
// 	optional string string_array=17;
// 	optional Tuple tuple_array=18;
// }

message NumpyNdArray {
	string array_type=1;
	repeated NumpyNdArray numpy_array=2;
	repeated bytes bytes_array=3;
	repeated double double_array=4;
	repeated float float_array=5;
	repeated int32 int32_array=6;
	repeated int64 int64_array=7;
	repeated uint32 uint32_array=8;
	repeated uint64 uint64_array=9;
	repeated sint32 sint32_array=10;
	repeated sint64 sint64_array=11;
	repeated fixed32 fixed32_array=12;
	repeated fixed64 fixed64_array=13;
	repeated sfixed32 sfixed32_array=14;
	repeated sfixed64 sfixed64_array=15;
	repeated bool bool_array=16;
	repeated string string_array=17;
}
//@Todo map<name, value>
message BentoServiceMessage {
	oneof io_descriptor{
		string text=2;
		NumpyNdArray numpy_ndarray = 3;
	}
	

	// PandasDataframe dataframe = 4;
	// ImageFile image = 5;
}

message BentoServiceInput {
	BentoServiceMessage input = 1;
	//api_name
}

message BentoServiceOutput {
	BentoServiceMessage output= 1;
	//status
	//error
}

service BentoML {
	rpc api_func(BentoServiceInput) returns (BentoServiceOutput);
}
